File: src/parsers/gemini.ts
/**
 * parsers/gemini.ts — Gemini CLI sessions.
 *
 * Store: ~/.gemini/tmp/<slug>/chats/session-*.{json,jsonl}. Two formats coexist:
 *   OLD .json   one object { sessionId, startTime, lastUpdated, messages:[…], summary? }
 *   NEW .jsonl  a change log: line 0 header { sessionId, startTime, projectHash, kind };
 *               then { "$set": { messages:[…] } } snapshot; then per-message
 *               append records { id, timestamp, type, content, model?, thoughts?,
 *               toolCalls?, tokens? }. Folded on read.
 * A sessionId can span multiple chunk files → merge by sessionId, order by
 * startTime, dedup messages by id. The synthetic first turn (user message whose
 * text opens with "<session_context>") is filtered out.
 *
 * Event mapping:
 *   type=user            → user
 *   type=gemini          → assistant text (usage from .tokens.total); .thoughts[]
 *                          → thinking; .toolCalls[] → tool_call (+ tool_result)
 *   type=info | error    → system (subtype = the native type)
 *
 * cwd not in transcript → recovered from <slug>/.project_root or projects.json.
 * No native title; old .json may carry a summary. Tree A via message id (no
 * parent pointer in the format). No cross-session fork edge.
 */

import { readdirSync, readFileSync, statSync } from "fs";
import { join } from "path";
import { GEMINI_PROJECTS_JSON, GEMINI_TMP } from "../paths.js";
import type { Event, Parser, UnifiedSession } from "../types.js";
import {
  asDict,
  asNum,
  asStr,
  compact,
  EventBuilder,
  fileSource,
  flattenContent,
  isoToEpochSec,
  parseJsonLine,
  readJson,
} from "../util.js";

interface GMsg {
  id?: string;
  timestamp?: string;
  type?: string; // user | gemini | info | error
  content?: unknown;
  message?: unknown;
  model?: unknown;
  thoughts?: unknown;
  toolCalls?: unknown;
  tokens?: unknown;
}

interface Chunk {
  path: string;
  sessionId: string | null;
  startTime: string | null;
  lastUpdated: string | null;
  summary: string | null;
  projectHash: string | null;
  slug: string;
  messages: GMsg[];
}

/** Fold one file (old .json or new .jsonl) into a chunk. */
function readChunk(path: string, slug: string): Chunk | null {
  const c: Chunk = {
    path,
    sessionId: null,
    startTime: null,
    lastUpdated: null,
    summary: null,
    projectHash: null,
    slug,
    messages: [],
  };
  if (path.endsWith(".jsonl")) {
    let text: string;
    try {
      text = readFileSync(path, "utf8");
    } catch {
      return null;
    }
    for (const line of text.split("\n")) {
      const r = parseJsonLine(line);
      if (!r) continue;
      if (r.sessionId) {
        c.sessionId ??= asStr(r.sessionId);
        c.startTime ??= asStr(r.startTime);
        c.projectHash ??= asStr(r.projectHash);
        c.lastUpdated = asStr(r.lastUpdated) ?? c.lastUpdated;
        continue;
      }
      if (r.$set && typeof r.$set === "object") {
        const set = r.$set as Record<string, unknown>;
        if (Array.isArray(set.messages)) c.messages.push(...(set.messages as GMsg[]));
        continue;
      }
      if (r.type && (r.type === "user" || r.type === "gemini" || r.type === "info" || r.type === "error")) {
        c.messages.push(r as GMsg);
      }
    }
  } else {
    const obj = asDict(readJson(path));
    if (!obj) return null;
    c.sessionId = asStr(obj.sessionId);
    c.startTime = asStr(obj.startTime);
    c.lastUpdated = asStr(obj.lastUpdated);
    c.summary = asStr(obj.summary);
    c.projectHash = asStr(obj.projectHash);
    if (Array.isArray(obj.messages)) c.messages.push(...(obj.messages as GMsg[]));
  }
  return c.sessionId ? c : null;
}

/** All (slug, chatfile) pairs under ~/.gemini/tmp. */
function listChunkFiles(): { path: string; slug: string }[] {
  let slugs: string[];
  try {
    slugs = readdirSync(GEMINI_TMP);
  } catch {
    return [];
  }
  const out: { path: string; slug: string }[] = [];
  for (const slug of slugs) {
    const chatsDir = join(GEMINI_TMP, slug, "chats");
    let entries: string[];
    try {
      if (!statSync(chatsDir).isDirectory()) continue;
      entries = readdirSync(chatsDir);
    } catch {
      continue;
    }
    for (const e of entries) {
      if (e.startsWith("session-") && (e.endsWith(".json") || e.endsWith(".jsonl")))
        out.push({ path: join(chatsDir, e), slug });
    }
  }
  return out;
}

/** Group chunks by sessionId. */
function bySession(): Map<string, Chunk[]> {
  const map = new Map<string, Chunk[]>();
  for (const { path, slug } of listChunkFiles()) {
    const c = readChunk(path, slug);
    if (!c || !c.sessionId) continue;
    const arr = map.get(c.sessionId);
    if (arr) arr.push(c);
    else map.set(c.sessionId, [c]);
  }
  return map;
}

function resolveCwd(slug: string): string | null {
  try {
    const t = readFileSync(join(GEMINI_TMP, slug, ".project_root"), "utf8").trim();
    if (t) return t;
  } catch {
    /* ignore */
  }
  const projects = asDict(asDict(readJson(GEMINI_PROJECTS_JSON)).projects);
  for (const [dir, s] of Object.entries(projects)) if (s === slug) return dir;
  return null;
}

function isSynthetic(m: GMsg): boolean {
  if (m.type !== "user") return false;
  return flattenContent(m.content ?? m.message).startsWith("<session_context>");
}

/** Merge + order + dedup a session's chunks into one message list. */
function mergeMessages(chunks: Chunk[]): GMsg[] {
  const sorted = [...chunks].sort((a, b) => (a.startTime ?? "").localeCompare(b.startTime ?? ""));
  const seen = new Set<string>();
  const merged: GMsg[] = [];
  for (const c of sorted) {
    for (const m of c.messages) {
      if (m.id && seen.has(m.id)) continue;
      if (m.id) seen.add(m.id);
      merged.push(m);
    }
  }
  merged.sort((a, b) => (a.timestamp ?? "").localeCompare(b.timestamp ?? ""));
  return merged;
}

export const geminiParser: Parser = {
  agent: "gemini",

  listSessions(sinceEpochSec) {
    const out: UnifiedSession[] = [];
    for (const [sessionId, chunks] of bySession()) {
      const merged = mergeMessages(chunks);
      const first = chunks.reduce((a, b) => ((a.startTime ?? "") <= (b.startTime ?? "") ? a : b));
      const last = chunks.reduce((a, b) => ((a.lastUpdated ?? "") >= (b.lastUpdated ?? "") ? a : b));
      const startedAt = isoToEpochSec(first.startTime);
      const endedAt = isoToEpochSec(last.lastUpdated) ?? startedAt;
      if (sinceEpochSec != null && (endedAt ?? 0) < sinceEpochSec) continue;
      let turns = 0;
      let model: string | null = null;
      for (const m of merged) {
        if ((m.type === "user" && !isSynthetic(m)) || m.type === "gemini") turns++;
        if (m.type === "gemini") model ??= asStr(m.model);
      }
      const summary = chunks.map((c) => c.summary).find((s) => s) ?? null;
      out.push({
        id: `gemini:${sessionId}`,
        agent: "gemini",
        cwd: resolveCwd(first.slug),
        startedAt,
        endedAt,
        title: null,
        model,
        provider: "google",
        gitBranch: null,
        parentSessionId: null,
        forkedFrom: null,
        turnCount: turns,
        eventCount: turns,
        sources: chunks.map((c) => fileSource(c.path, c.path.endsWith(".jsonl") ? "jsonl" : "json")),
        fingerprint: `${merged.length}:${endedAt ?? 0}`,
        meta: { summary, projectHash: first.projectHash, sourceFileCount: chunks.length },
      });
    }
    return out;
  },

  readEvents(nativeId): Event[] {
    const chunks = bySession().get(nativeId);
    if (!chunks) return [];
    const merged = mergeMessages(chunks);
    const out = new EventBuilder();
    for (const m of merged) {
      if (isSynthetic(m)) continue;
      const ts = isoToEpochSec(m.timestamp);
      const type = m.type;
      const id = asStr(m.id);
      if (type === "user") {
        out.push("user", flattenContent(m.content ?? m.message), ts, { nativeId: id });
      } else if (type === "gemini") {
        const tok = asNum(asDict(m.tokens).total);
        out.push("assistant", flattenContent(m.content ?? m.message), ts, { nativeId: id, tokens: tok });
        if (Array.isArray(m.thoughts)) {
          for (const t of m.thoughts) {
            const td = asDict(t);
            const txt = [asStr(td.subject), asStr(td.description)].filter(Boolean).join(": ");
            out.push("thinking", txt || compact(t), ts, { nativeId: id });
          }
        }
        if (Array.isArray(m.toolCalls)) {
          for (const tc of m.toolCalls) {
            const t = asDict(tc);
            const name = asStr(t.name);
            const callId = asStr(t.callId) ?? asStr(t.id);
            out.push("tool_call", compact(t.args), ts, { name, callId, nativeId: id });
            const result = t.resultDisplay ?? t.result;
            if (result != null)
              out.push("tool_result", compact(result), ts, {
                name,
                callId,
                nativeId: id,
                isError: asStr(t.status) === "error" || t.success === false,
              });
          }
        }
      } else if (type === "info" || type === "error") {
        out.push("system", flattenContent(m.content ?? m.message), ts, { subtype: type, nativeId: id });
      }
    }
    return out.events;
  },
};

File: src/parsers/gemini.rs
use super::{Parser, SessionCandidate};
use crate::model::{
    compact, flatten, Agent, Capture, Event, EventKind, NativeSource, ParsedSession, Session,
};
use anyhow::Result;
use chrono::DateTime;
use serde_json::{json, Value};
use std::{fs, path::Path};
use walkdir::WalkDir;

pub struct GeminiParser;
fn s(v: Option<&Value>) -> Option<String> {
    v.and_then(Value::as_str).map(str::to_owned)
}
fn ts(v: Option<&Value>) -> Option<i64> {
    v.and_then(Value::as_str)
        .and_then(|x| DateTime::parse_from_rfc3339(x).ok())
        .map(|x| x.timestamp_millis())
}
fn parse(path: &Path, root: &Path, candidate: &SessionCandidate) -> Result<ParsedSession> {
    let text = fs::read_to_string(path)?;
    let mut rows = Vec::new();
    let mut sid = None;
    let mut start = None;
    let mut end = None;
    let mut model = None;
    for line in text.lines() {
        let Ok(v): Result<Value, _> = serde_json::from_str(line) else {
            continue;
        };
        if let Some(x) = s(v.get("sessionId")) {
            sid = Some(x);
            start = start.or(ts(v.get("startTime")));
            end = ts(v.get("lastUpdated")).or(end);
            // Gemini's older store is one JSON object with an embedded
            // `messages` array rather than a change log.
            if let Some(messages) = v.get("messages").and_then(Value::as_array) {
                rows.extend(messages.iter().cloned());
            }
            continue;
        }
        if let Some(set) = v.get("$set") {
            if let Some(ms) = set.get("messages").and_then(Value::as_array) {
                rows.extend(ms.iter().cloned());
            }
            continue;
        }
        if matches!(
            v.get("type").and_then(Value::as_str),
            Some("user") | Some("gemini") | Some("info") | Some("error")
        ) {
            rows.push(v);
        }
    }
    let sid = sid.unwrap_or_else(|| {
        path.file_stem()
            .and_then(|x| x.to_str())
            .unwrap_or("unknown")
            .into()
    });
    let mut events = Vec::new();
    let mut seen = std::collections::HashSet::new();
    for r in rows {
        let typ = r.get("type").and_then(Value::as_str).unwrap_or("");
        let id = s(r.get("id"));
        if !seen.insert(id.clone().unwrap_or_default()) {
            continue;
        }
        let t = ts(r.get("timestamp"));
        if start.is_none() {
            start = t;
        }
        if t.is_some() {
            end = t;
        }
        match typ {
            "user" => {
                let txt = flatten(
                    r.get("content")
                        .or_else(|| r.get("message"))
                        .unwrap_or(&Value::Null),
                );
                if !txt.starts_with("<session_context>") {
                    let mut e = Event::new(EventKind::User, txt);
                    e.native_id = id;
                    e.created_at_ms = t;
                    events.push(e)
                }
            }
            "gemini" => {
                let mut e = Event::new(
                    EventKind::Assistant,
                    flatten(
                        r.get("content")
                            .or_else(|| r.get("message"))
                            .unwrap_or(&Value::Null),
                    ),
                );
                e.native_id = id.clone();
                e.created_at_ms = t;
                e.model = s(r.get("model"));
                model = model.or(e.model.clone());
                events.push(e);
                if let Some(arr) = r.get("thoughts").and_then(Value::as_array) {
                    for x in arr {
                        let mut q = Event::new(EventKind::Thinking, compact(x));
                        q.native_id = id.clone();
                        q.created_at_ms = t;
                        events.push(q)
                    }
                }
                if let Some(arr) = r.get("toolCalls").and_then(Value::as_array) {
                    for x in arr {
                        let mut q =
                            Event::new(EventKind::ToolCall, compact(x.get("args").unwrap_or(x)));
                        q.name = s(x.get("name"));
                        q.call_id = s(x.get("callId").or_else(|| x.get("id")));
                        q.native_id = id.clone();
                        q.created_at_ms = t;
                        let qname = q.name.clone();
                        let qcall = q.call_id.clone();
                        events.push(q);
                        if x.get("result").is_some() {
                            let mut z = Event::new(
                                EventKind::ToolResult,
                                compact(x.get("result").unwrap()),
                            );
                            z.call_id = qcall;
                            z.name = qname;
                            z.native_id = id.clone();
                            z.created_at_ms = t;
                            events.push(z)
                        }
                    }
                }
            }
            "info" | "error" => {
                let mut e = Event::new(
                    EventKind::System,
                    flatten(
                        r.get("content")
                            .or_else(|| r.get("message"))
                            .unwrap_or(&Value::Null),
                    ),
                );
                e.subtype = Some(typ.into());
                e.native_id = id;
                e.created_at_ms = t;
                events.push(e)
            }
            _ => {}
        }
    }
    let restore = path
        .strip_prefix(root)
        .ok()
        .and_then(|x| x.to_str())
        .unwrap_or_else(|| {
            path.file_name()
                .and_then(|x| x.to_str())
                .unwrap_or("session.jsonl")
        })
        .replace(std::path::MAIN_SEPARATOR, "/");
    let source = NativeSource {
        locator: path.display().to_string(),
        kind: "jsonl".into(),
        restore_path: restore,
        role: None,
        bytes: candidate.bytes,
        mtime_ns: candidate.mtime_ns,
        mode: None,
        capture: Some(Capture::File {
            path: path.display().to_string(),
        }),
    };
    Ok(ParsedSession {
        session: Session {
            id: format!("gemini:{sid}"),
            agent: Agent::Gemini,
            cwd: None,
            started_at_ms: start,
            ended_at_ms: end,
            title: None,
            model,
            provider: Some("google".into()),
            git_branch: None,
            parent_session_id: None,
            forked_from: None,
            meta: json!({}),
            fingerprint: format!("{}:{}", events.len(), end.unwrap_or_default()),
            sources: vec![source],
        },
        events,
    })
}
impl Parser for GeminiParser {
    fn agent(&self) -> Agent {
        Agent::Gemini
    }
    fn discover(&self, root: &Path) -> Result<Vec<SessionCandidate>> {
        let mut out = Vec::new();
        if !root.exists() {
            return Ok(out);
        };
        for e in WalkDir::new(root)
            .follow_links(false)
            .into_iter()
            .filter_map(Result::ok)
        {
            if e.file_type().is_file() && e.file_name().to_string_lossy().starts_with("session-") {
                if let Ok(candidate) = SessionCandidate::file(e.path().to_path_buf()) {
                    out.push(candidate);
                }
            }
        }
        Ok(out)
    }

    fn parse(&self, candidate: &SessionCandidate, root: &Path) -> Result<Option<ParsedSession>> {
        Ok(Some(parse(&candidate.path, root, candidate)?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn reads_legacy_json_message_array() {
        let dir = tempdir().unwrap();
        let path = dir.path().join("session-legacy.json");
        fs::write(
            &path,
            serde_json::json!({
                "sessionId": "g-1",
                "startTime": "2026-08-19T00:00:00Z",
                "lastUpdated": "2026-08-19T00:00:02Z",
                "messages": [
                    {"id":"u","type":"user","content":"hello"},
                    {"id":"a","type":"gemini","content":"world"}
                ]
            })
            .to_string(),
        )
        .unwrap();
        let parsed = GeminiParser.parse_all(dir.path()).unwrap();
        assert_eq!(parsed.len(), 1);
        assert_eq!(
            parsed[0]
                .events
                .iter()
                .filter(|e| e.kind == EventKind::User)
                .count(),
            1
        );
        assert_eq!(
            parsed[0]
                .events
                .iter()
                .filter(|e| e.kind == EventKind::Assistant)
                .count(),
            1
        );
    }
}
